docs/src/adr/024adr-arena-pass.md

```diff
@@ -0,0 +1,72 @@
+# ADR-024: Arena computation isolation
+
+| authors                                | last revised    |
+| -------------------------------------- | --------------: |
+| Jure Kukovec                           | 2023-03-01      |
+
+**Table of Contents**
+
+- [Summary (Overview)](#summary)
+- [Context (Problem)](#context)
+- [Options (Alternatives)](#options)
+- [Solution (Decision)](#solution)
+- [Consequences (Retrospection)](#consequences)
+
+## Summary
+
+<!-- Statement to summarize, following the following formula: -->
+
+In the context of addressing tech debt\
+facing increasing difficulties understanding and modifying the BMC pass\
+we decided to decouple arena construction from rewriting rules\
+to achieve better modularity, readability, and maintainability \
+accepting a reasonable time investment into refactoring.\
+
+## Context
+
+<!-- Communicates the forces at play (technical, political, social, project).
+     This is the story explaining the problem we are looking to resolve.
+-->
+
+The core of the model checking pass is tangled and messy; due to its ad-hoc development in the past, classes and dependencies were added incrementally, often in arbitrary, suboptimal ways.
```

Comment:
I would prefer to have a more objective description of what is happening in the rewriting rules. As we discussed, having a diagram for one rule would really help to untangle the call patterns in the current implementation. If you don't like diagrams, the same could be achieved by decorating a single rule with comments that explain what is going on and why it is not optimal.